### ISSUE TITLE
Inefficient memory usage in message passing

### ISSUE THREAD

**Issue by dar326** [bayespy/bayespy#83]
Hi,

I'm attempting to use BayesPy with a somewhat large hierarchical model. My model has a structure including 10 nodes (9 Categorial, 1 Multinomial). Each of the Categorical nodes has an unknown Dirichlet prior (theta) and all but one (X_1) has a parent represented by one of the other Categorical nodes. One of the Categorical nodes (X_9) is the parent of the Multinomial node (Y). The Multinomial node is observed and has 256 possible outcomes. For each of ~4000 possible outcomes of X_9, we have a known distribution over the 256 outcomes of Y. The observations for Y consist of a variable number of trials N_i. For simplicity, I am restricting the number of observations to 1000. So, we have obs_0, obs_1, ..., obs_S with N_0, N_1,...,N_S trials, respectively. Currently, N_i is ~100. These nodes have been implemented as mixture distributions when the node has a parent. The one node without a parent (X_1) is a Categorical node.

When setting up this model with BayesPy using the VB inference engine, all of the data and the model require less than 2 GB of memory. However, there is a huge growth in memory when I observe data and then update the model. In particular, the "update" line below (when using memory profiling tools and observing memory usage with the "top" command), causes a huge spike in memory usage (maxing out above ~100 GB):

```
Q = VB(*(thetas + Xs + [Y]), tol=1e-4)
Q.update(repeat=1000, tol=1e-4)
```

The number of observations (1000 in this example) is just a small sample of my total number of observations. I can deal with sub-sampling my data using stochastic variational inference to learn my priors. So, in this case, while the memory usage is not ideal, I can run the model.

However, this is just a simplified version of my real problem. The real problem has ~65k possible outcomes for the Multinomial node. Even with a sample of 1000 observations, I run into memory errors which I have tracked to the model update function. Specifically, the traceback includes the following code:

```
self = <bayespy.inference.vmp.nodes.multinomial.MultinomialDistribution object at 0x7f7c5c78cda0>
u = [array([[[ 0.,  0.,  0., ...,  0.,  0.,  0.]],

       [[ 0.,  0.,  0., ...,  0.,  0.,  0.]],

       [[ 0.,  0.,  0.,..., ...,  0.,  0.,  0.]],

       [[ 0.,  0.,  0., ...,  0.,  0.,  0.]],

       [[ 0.,  0.,  0., ...,  0.,  0.,  0.]]])]
phi = [array([[[-11.09035492, -11.09035492, -11.09035492, ..., -11.09035492,
         -11.09035492, -11.09035492],
        [...  [-11.09035492, -11.09035492, -11.09035492, ..., -11.09035492,
         -11.09035492, -11.09035492]]], dtype=float32)]
g = array([0]), f = 0, ndims = [1]

    def compute_logpdf(self, u, phi, g, f, ndims):
        """ Compute E[log p(X)] given E[u], E[phi], E[g] and
            E[f]. Does not sum over plates."""
    
        # TODO/FIXME: Should I take into account what is latent or
        # observed, or what is even totally ignored (by the mask).
        L = g + f
        for (phi_i, u_i, ndims_i) in zip(phi, u, ndims):
            # Axes to sum (dimensions of the variable, not the plates)
            axis_sum = tuple(range(-ndims_i,0))
            # Compute the term
            # TODO/FIXME: Use einsum!
>           L = L + np.sum(phi_i * u_i, axis=axis_sum)
E           MemoryError
```

So, it appears that this summation line when computing the log pdf for my Multinomial node is responsible for the large memory usage. I noticed that the comment above the line where the memory error occurs mentions using NumPy's einsum at some point in the future.

I'm wondering if the plan to use einsum is to reduce memory usage? If so, would you expect such a change to reduce the memory usage in my model or is this model simply too large for BayesPy?

Thanks,
Darryl

**Comment by jluttine** [comment]
Hi, thanks for the report! Good to have such performance related reports. I can think of three possible causes: 1) The model is just too big. 2) Some message calculations are implemented inefficiently with respect to memory usage. 3) There is some mistake in model definition, for instance, with respect to plates. Could you provide some code for constructing the model so I could take a look and play around a little bit?

**Comment by dar326** [comment]
No problem! I have included a file ("test_model.npz") which can be used to build an example model based on the code below. I really hope the answer is not due to cause (1) but, hopefully, by playing around with the model a little, you can let me know which of the possible causes is at issue here.

Thanks!
Darryl

[test_model.npz.zip](https://github.com/bayespy/bayespy/files/395227/test_model.npz.zip)



```
import numpy as np
from bayespy import nodes
from bayespy.inference import VB

model_data = np.load("test_model.npz")
thetas = []
Xs = []
obs = model_data["sample"]
thetas.append(nodes.Dirichlet(model_data["theta_0"]))
Xs.append(nodes.Categorical(thetas[-1], plates=(obs.shape[0],))
for i in range(1, 9):
    thetas.append(nodes.Dirichlet(model_data["theta_%s" % i]))
    Xs.append(nodes.Mixture(Xs[-1], nodes.Categorical, thetas[-1], cluster_plate=-2))

Y = nodes.Mixture(Xs[-1], nodes.Multinomial, obs.sum(axis=1), model_data["k_dist"], cluster_plate=-2)
Y.observe(obs)
Q = VB(*(thetas + Xs + [Y]))
Q.update(repeat=1000, tol=1e-4)
```

**Comment by dar326** [comment]
Hi Jaakko,

Just wondering if you had gotten a chance to look at the sample model that I sent previously to see what might be the cause of the performance issues that I am experiencing?

Thanks,
Darryl

**Comment by jluttine** [comment]
Not yet, sorry. Been very busy with other matters. I'll try to take a look as soon as possible. Sorry for the delay.

**Comment by jluttine** [comment]
Sorry for the exceptionally huge delay in my responses.. I've been extremely busy lately for various reasons. I took a look and want to inform about my findings although I still didn't fix it: the reason is (2), that is, BayesPy computes specific messages inefficiently with respect to memory usage. In particular, in your example, when BayesPy computes the message from the Mixture node `Y` to the Dirichlet node `theta[8]`, `compute_message_to_parent` computes an intermediate array of size `(1782, 1000, 3853)`, which is... quite... large... uh...

Note to self: the issue is on this line:
https://github.com/bayespy/bayespy/blob/184fe123d1191ef0e52fa5fc4b629ac1fce65be0/bayespy/inference/vmp/nodes/mixture.py#L146
Need to do the summation at the same time using einsum.

I've been having plans to start using Tensorflow (or Theano, or autograd) for computing these gradient messages automatically and thus probably avoiding these issues elegantly. Until I have time to implement Tensorflow support, I need to fix this.

I'll try to reply to you faster and overall handle this issue faster now. I'll take a look on how I could easily solve this issue. And big thanks for the very good issue report!

**Comment by jluttine** [comment]
I'll write a few notes for myself here so it's all in one place. You may ignore this message.

The current message passing functions don't allow writing `compute_message_to_parent` function efficiently because the mask is applied *after* the computation, so it cannot sum anything - and it even cannot know what axes it should sum. Thus, einsum cannot currently be used to obtain lower memory usage. The issue is general and applies to other nodes as well, although the issue has effect only on such nodes in which some axes of the parents are summed over (e.g., Mixture, Dot)

The message passing system needs to be slightly changed. `compute_message_to_parent` function should receive the mask and final shape so it can compute the final result without computing huge temporary result arrays. If we think in terms of gradient computations, the log probability density term is naively written `sum(mask * L)` where `L` contains the pdf terms for each plate. Thus, when computing the gradient with backpropagation, `mask` should be applied at the beginning before going inside `L`, not after it, because that would require that all the necessary axes need to be kept until mask is applied.

**Comment by dar326** [comment]
Thanks, Jaakko! Glad I could make a contribution.

**Comment by jluttine** [comment]
Hi! I made a "fix" to BayesPy which should remove these memory errors in your case. I needed to revamp the whole gradient messaging and I did it only for a few nodes so I think only Dirichlet, Categorical and Mixture work, all other nodes are broken. This state is available in branch `feature/backpropagation-messaging`. To install it, this should work: `pip install git+https://github.com/bayespy/bayespy.git@backpropagation-messaging`.

If you happen to try it and experience any issues, please report so I can fix those for you. Disclaimer: I implemented it quite quickly, so there may be bugs.

**Comment by dar326** [comment]
Thank you! I will give it a try and report back if I run into any problems.

**Comment by dar326** [comment]
Just tried to install the new branch but getting an error:

```$ pip install git+https://github.com/bayespy/bayespy.git@backpropagation-messaging
Collecting git+https://github.com/bayespy/bayespy.git@backpropagation-messaging
  Cloning https://github.com/bayespy/bayespy.git (to backpropagation-messaging) to /tmp/pip-qc5xgelt-build
  Could not find a tag or branch 'backpropagation-messaging', assuming commit.
error: pathspec 'backpropagation-messaging' did not match any file(s) known to git.
Command "git checkout -q backpropagation-messaging" failed with error code 1 in /tmp/pip-qc5xgelt-build```

Also, poked around the github repo and don't see any reference to the "backpropagation-messaging." Perhaps the branch wasn't checked in?

**Comment by jluttine** [comment]
Sorry! I accidentally had pushed it my personal repo at jluttine/bayespy. It should now be available here. Sorry about this.

**Comment by dar326** [comment]
No problem. Thanks for the fix. Just a note that I had to enter the following command to install the branch:
```pip install git+https://github.com/bayespy/bayespy.git@feature/backpropagation-messaging```

**Comment by dar326** [comment]
So, no longer getting the error that I reported before but still getting a memory error in a different section of the code base:

```
self = <bayespy.inference.vmp.nodes.multinomial.MultinomialDistribution object at 0x7f0204a83cc0>
x = array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ..., 
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=uint32)
mask = True

    def compute_fixed_moments_and_f(self, x, mask=True):
        r"""
            Compute the moments and :math:`f(x)` for a fixed value.
            """
    
        # Check that counts are valid
        x = np.asanyarray(x)
        if not misc.isinteger(x):
            raise ValueError("Counts must be integers")
        if np.any(x < 0):
            raise ValueError("Counts must be non-negative")
        if np.any(np.sum(x, axis=-1) != self.N):
            raise ValueError("Counts must sum to the number of trials")
    
        # Moments is just the counts vector
        u0 = x.copy()
        u = [u0]
    
>       f = special.gammaln(self.N+1) - np.sum(special.gammaln(x+1), axis=-1)
E       MemoryError
```

Looks like the error is happening again on the np.sum call. Any chance this issue has a similar fix?

**Comment by jluttine** [comment]
That is a bit weird place for a memory error.. All the variables should have the same shape as x or smaller. Of course, if x is so large that you can only fit a few arrays of that size into your memory, then this definitely may happen. Can you easily find out what are the shapes of x and self.N in your function call when getting the memory error (e.g., with debugger or by inserting `print(np.shape(self.N), np.shape(x))` into that method)?

**Comment by dar326** [comment]
Please ignore my last error report. Moving the analysis to a server with more memory results in the code running without any memory errors. Looks like the fix worked. Thanks!

**Comment by jluttine** [comment]
Ok, glad to hear!

**Comment by jluttine** [comment]
Actually, I should keep this open until I've fixed the message passing in BayesPy, not just in some temporary branch..
